src/rust/ensogl/lib/components/src/file_browser/icons.rs

```diff
@@ -0,0 +1,487 @@
+// === Project ===
+// ===============
+
+/// The shape definition for a project icon.
+pub mod project {
+    use super::*;
+
+    ensogl_core::define_shape_system! {
+        (style:Style,color_rgba:Vector4,stroke_width:f32) {
+            let stroke_width : Var<Pixels> = stroke_width.into();
+
+            let left  = Rect((&stroke_width,10.0.px())).translate_x((-5.5).px()+&stroke_width/2.0);
+            let right = Rect((&stroke_width,10.0.px())).translate_x(5.5.px()-&stroke_width/2.0);
+
+            let top_ellipse = Ellipse(5.5.px(),1.5.px());
+            let top_upper   = &top_ellipse - top_ellipse.translate_y(-&stroke_width);
+            let top_lower   = &top_ellipse - top_ellipse.translate_y(&stroke_width);
+            let top         = top_upper + top_lower;
+            let top         = top.translate_y(5.0.px());
+
+            let bottom_outer = Ellipse(5.5.px(),2.0.px());
+            let bottom_outer = bottom_outer.translate_y(0.5.px()-&stroke_width);
+            let bottom_inner = Ellipse(4.5.px(),1.5.px());
+            let bottom       = bottom_outer - bottom_inner;
+            let bottom       = bottom * HalfPlane();
+            let bottom       = bottom.translate_y((-5.5).px()+&stroke_width);
+
+            let upper_middle_outer = Ellipse(5.0.px(),1.6666.px());
+            let upper_middle_inner = upper_middle_outer.translate_y(&stroke_width/2.0);
+            let upper_middle       = upper_middle_outer - upper_middle_inner;
+            let upper_middle       = upper_middle.translate_y(1.9166.px());
+
+            let lower_middle_outer = Ellipse(5.0.px(),1.83333.px());
+            let lower_middle_inner = lower_middle_outer.translate_y(&stroke_width/2.0);
+            let lower_middle       = lower_middle_outer - lower_middle_inner;
+            let lower_middle       = lower_middle.translate_y((-1.4166).px());
+
+            let shape = left + right + top + bottom + upper_middle + lower_middle;
+            let shape = shape.fill(color_rgba);
+            let shape = shape.shrink(SHRINK_AMOUNT.px());
+            shape.into()
+        }
+    }
+}
+
+/// A project icon.
+#[derive(Debug)]
+pub struct Project(project::View);
+
+impl Project {
+    /// Construct a new icon.
+    pub fn new() -> Self {
+        let shape_view = project::View::new(Logger::new("file_browser::icon::Project"));
+        shape_view.size.set(Vector2(ICON_SIZE, ICON_SIZE));
+        let icon = Project(shape_view);
+        icon.set_focused(false);
+        icon
+    }
+}
+
+impl display::Object for Project {
+    fn display_object(&self) -> &Instance {
+        self.0.display_object()
+    }
+}
+
+impl DynamicIcon for Project {
+    fn set_focused(&self, focused:bool) {
+        self.0.stroke_width.set(if focused {1.5} else {1.0});
+        self.0.color_rgba.set(if focused {focused_color().into()} else {standard_color().into()})
+    }
+}
+
+impl Default for Project {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
This is very repetitive pattern. Do you think it is worth deduplicating? Perhaps through macro, or maybe even just the generic structure?

Reply:
I thought about it too, but I do not think that it is worth it to use a macro for this.

What do you mean by "the generic structure"?

Reply:
Actually, by generic structure I meant exactly a structure with generic parameters, like `struct Icon<View>` perhaps making some helper traits is not avoidable.

Reply:
> Why not macro? You have written " I do not think that it is worth it to use a macro for this." - but you have not explained why you think its not worth.

I have not explained it because it sounded like Adam is aware of some of the downsides and his comment was more a suggestion than a request for a discussion.

Some downsides of using macros are: Currently, it is obvious what types are defined and what traits are implemented in this file (apart from the `define_shape_system!` definitions). If we use macros then you have to understand the macro first, before you can make any sense of this file. Macro definitions provide few static correctness guarantees. If a macro fails then the error messages are typically quite difficult to understand. IntelliJ IDEA and VS Code with the popular Rust extension provide bad support for things that you pass to a macro. And finally, not every developer is familiar with macro definitions.